### PR TITLE
Auto chaptermarker & custom bookmark location

### PR DESCRIPTION
**Pull request by hypoflex**
#446
Once a chapter is previewed it will be marked as "last read"

Using the build in viewer/reader will keep track of where you left off. once you close the build in viewer/reader it will automatically put that last chapter as "last read".

### CODE REVIEW

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (hypoflex) [inline src/web/lib/hakuneko/frontend@classic-dark/chapters.html:102]: Not sure why its doing this...?
tried reforking the hole project and removing it from my machine.
but it keeps editing these, bug?!

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (hypoflex) [inline src/web/lib/hakuneko/frontend@classic-dark/chapters.html:587]: same story here, nothing changed nothing added...
there are more places it "fucks" with the code like this

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (hypoflex) [inline src/web/lib/hakuneko/frontend@classic-dark/chapters.html]: added listener to find if viewer is closed

**Review by hypoflex — COMMENTED** [review]
-

**Comment by hypoflex** [comment]
https://github.com/manga-download/hakuneko/pull/448/commits/fe98c0cb517e55683e700b8ac98039cbd755b8e9 Needs to be reviewed for approval as well for #447 this way the use can choose where to save the 2 files.
default location on default platform place so no further conflict with existing installations.

File: src/web/lib/hakuneko/frontend@classic-dark/pages.html
Comment (27136773) [inline src/web/lib/hakuneko/frontend@classic-dark/pages.html]: May just set the chapter marker directly:
```javascript
Engine.ChaptermarkManager.addChaptermark( this.selectedChapter );
```
Then no changes in `chapters.html` are required at all

File: src/web/lib/hakuneko/frontend@classic-light/pages.html
Comment (27136773) [inline src/web/lib/hakuneko/frontend@classic-light/pages.html]: May just set the chapter marker directly:
```javascript
Engine.ChaptermarkManager.addChaptermark( this.selectedChapter );
```
Then no changes in `chapters.html` are required at all

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (27136773) [inline src/web/lib/hakuneko/frontend@classic-dark/chapters.html:94]: No need to change this file at all when doing
https://github.com/manga-download/hakuneko/pull/448/files#r329122054

File: src/web/lib/hakuneko/frontend@classic-light/chapters.html
Comment (27136773) [inline src/web/lib/hakuneko/frontend@classic-light/chapters.html:94]: No need to change this file at all when doing
https://github.com/manga-download/hakuneko/pull/448/files#r329122332

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: Move method either to `Storage` or to the following both: `BookmarkManager` and `ChaptermarkManager`

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: Move method either to `Storage` or to the following both: `BookmarkManager` and `ChaptermarkManager`

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: Will `\\` path separator work on linux and macOS?

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: Will `\\` path separator work on linux and macOS?

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: Use `fs-extra` instead, it provides everything from `fs` but returns more convenient promises instead of callback functions

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: Moved to `Storage`

Reason for moving  to storage instead of keeping it in  settings is to minimize app performance impact?

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: i knew this would  maybe  cause an issue
have not tested on linux/osx yet, didnt have my laptop in reach, will test later

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: Reason for moving  to storage instead of keeping it in  settings is to minimize app performance impact?

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Settings.mjs:79]: This cant be tested well running  in npm. when clicking the browser icon both on this new directory browser as the already existing one for manga's it returns a {Promise Object} in the field, how can i  get around this.

right  now ive been just  switching the value within the code to test.

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: the `fs` function i grabbed from the `storage` file, where `fs` is still used, maybe good to change it there to `fs-extra` as well?

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: i knew this would maybe cause an issue
have not tested on linux/osx yet, didnt have my laptop in reach, will test later

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (hypoflex) [inline src/web/lib/hakuneko/frontend@classic-dark/chapters.html:439]: this line still "needed" though. its used to when preview is clicked  the chapter mark is updated.
So for example:
 when opening Tokyo ghoul and having no  chaptermark, and starting at chapter 100
the marker will  be  set there.

on the other hand i could see confusion happening with this if a user desides to go back a chapter the marker also gets updated.

Comment (hypoflex) [reply src/web/lib/hakuneko/frontend@classic-dark/chapters.html:439]: Does not seem to work on OSX.

will first convert other changes before testing on other platform again

File: src/web/lib/hakuneko/frontend@classic-dark/pages.html
Comment (hypoflex) [reply src/web/lib/hakuneko/frontend@classic-dark/pages.html]: Changes made

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: Moved to `Storage`

**Comment by hypoflex** [comment]
Going to test changes on OSX now

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (27136773) [reply src/web/lib/hakuneko/frontend@classic-dark/chapters.html:102]: Removing the trailing space seems correct to me, this is in compliance with the current lint configuration.
Unfortunately linting is not yet supported for HTML files, so thats the reason the trailing whitespace was not detected as error.
Your IDE may see the lint configuration and adjust the errors automatically according to the configuration, which is correct.

File: src/web/lib/hakuneko/frontend@classic-dark/chapters.html
Comment (27136773) [reply src/web/lib/hakuneko/frontend@classic-dark/chapters.html:587]: Your change is correct, trailing white-spaces must be removed according to the lint configuration.

File: src/web/lib/hakuneko/frontend@classic-light/pages.html
Comment (27136773) [inline src/web/lib/hakuneko/frontend@classic-light/pages.html:330]: It is up to you to decide which actions should update the chapter marker, the following candiates may be possible:

- `showViewer`
- `hideViewer`
- `requestChapterUp`
- `requestChapterDown`

I think your current solution is fine, chapter mark will be updated when the reader is closed (which probably always is the case, except the program is closed while reading or crashes for some reason)

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: I think storage provides `path` or `this.path` so your separator issue can now be solved with `path.join(Engine.Settings.bookmarkDirectory.value, 'hakuneko.' + key)`

Comment (hypoflex) [reply src/web/mjs/engine/Storage.mjs]: `\\hakuneko` is indeed windows only, since windows needs a backslash and linux/osx needs a forward slash.

i was working on a switch case based on variable

```javascript
let platform = undefined;
switch(this.platform.indexOf){
case 'win':
 platform = '\\hakuneko'
 break;
case 'darwin':
 platform = '/hakuneko'
break;
}
```
(above code does not work as intended yet)

i will try what you posted, see what that results in

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [reply src/web/mjs/engine/Settings.mjs:79]: I forgot to fix the bug in dark theme.
Done now.
https://github.com/manga-download/hakuneko/commit/618ecde29a464e6a838fe03fb6c04738317f11d7

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: I think storage provides `path` or `this.path` so your separator issue can now be solved with `path.join(Engine.Settings.bookmarkDirectory.value, 'hakuneko.' + key)`

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [reply src/web/mjs/engine/Settings.mjs]: Yes, good idea

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [reply src/web/mjs/engine/Settings.mjs]: No its more to honor clean code principle (class responsibility), settings shall not handle bookmark and chapter-mark I/O.

**Comment by hypoflex** [comment]
- edited `pages.html` to nolonger require `chapter.html`
- changed `fs` to `fs-extra`
- moved `loadbookmark` and `savebookmark` to `Storage` instead of `Settings`
- changed the `Writefile` function to use `path.join` so separator works on all systems
-  folder async fixed by ronny https://github.com/manga-download/hakuneko/commit/618ecde29a464e6a838fe03fb6c04738317f11d7

File: src/web/mjs/engine/BookmarkManager.mjs
Comment (27136773) [inline src/web/mjs/engine/BookmarkManager.mjs:149]: Will bookmarks be saved after changing the directory in settings?
If not they may be lost when the directory is changed and the application is closed right after.

If not it might be a good idea to save the bookmarks whenever the settings changed event is fired
Example: https://github.com/manga-download/hakuneko/blob/618ecde29a464e6a838fe03fb6c04738317f11d7/src/web/mjs/connectors/NineAnime.mjs#L46

File: src/web/mjs/engine/ChaptermarkManager.mjs
Comment (27136773) [inline src/web/mjs/engine/ChaptermarkManager.mjs:40]: Will chaptermarks be saved after changing the directory in settings?
If not they may be lost when the directory is changed and the application is closed right after.

If not it might be a good idea to save the chaptermarks whenever the settings changed event is fired
Example: https://github.com/manga-download/hakuneko/blob/618ecde29a464e6a838fe03fb6c04738317f11d7/src/web/mjs/connectors/NineAnime.mjs#L46

File: src/web/mjs/engine/BookmarkManager.mjs
Comment (hypoflex) [reply src/web/mjs/engine/BookmarkManager.mjs:149]: good point!
looking into it

File: src/web/mjs/engine/BookmarkManager.mjs
Comment (hypoflex) [inline src/web/mjs/engine/BookmarkManager.mjs]: removed the name tags, these were mainly used to quick search without knowning the exact function name.

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Settings.mjs:285]: I had to add a new event listener here. 
the `onSettingsChanged` is being created twice, once here `line 285` and above `line 255`.

`Line 255` is being called once hakuneko is started.
`(Line 233 in current master)`
https://github.com/manga-download/hakuneko/blob/9a52af9cf2605bdb922a61b323f3b06ddb968220/src/web/mjs/engine/Settings.mjs#L233
problem with that is, if I bind the `savebookmark` function to `onSettingsChanged`
it will save a empty bookmark, since its not loaded yet.

File: src/web/mjs/engine/ChaptermarkManager.mjs
Comment (hypoflex) [inline src/web/mjs/engine/ChaptermarkManager.mjs:12]: Created new listener `onSettingsSaved` instead of `onSettingsChanged`, read comment below

File: src/web/mjs/engine/BookmarkManager.mjs
Comment (hypoflex) [inline src/web/mjs/engine/BookmarkManager.mjs:17]: Created new listener onSettingsSaved instead of onSettingsChanged, read comment below

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Settings.mjs]: `saveProfile` if settings file is not found? resolves #451

File: src/web/mjs/engine/ChaptermarkManager.mjs
Comment (27136773) [reply src/web/mjs/engine/ChaptermarkManager.mjs:12]: Unfortunately i hadn't time yet to refactor `Settings.mjs` to the new event target system, so it seems you need to use the deprecated document based system to use your new event.
In that case you also need to declare a string for this event in [Enums.mjs](https://github.com/manga-download/hakuneko/blob/9a52af9cf2605bdb922a61b323f3b06ddb968220/src/web/mjs/engine/Enums.mjs#L8), where you have do define a string for `EventListener.onSettingsSaved`, otherwise this is `undefined` which may lead to problems.

File: src/web/mjs/engine/BookmarkManager.mjs
Comment (27136773) [reply src/web/mjs/engine/BookmarkManager.mjs:17]: Unfortunately i hadn't time yet to refactor `Settings.mjs` to the new event target system, so it seems you need to use the deprecated document based system to use your new event.
In that case you also need to declare a string for this event in [Enums.mjs](https://github.com/manga-download/hakuneko/blob/9a52af9cf2605bdb922a61b323f3b06ddb968220/src/web/mjs/engine/Enums.mjs#L8), where you have do define a string for `EventListener.onSettingsSaved`, otherwise this is `undefined` which may lead to problems.

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: You may place this at the top and then change the `app` assignment to use this variable instead to avoid double `require` statements
```javascript
let electron = require( 'electron' );
let app = electron.remote.app;
```

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: use `this.app` instead of `electron.remote.app`

Comment (27136773) [reply src/web/mjs/engine/Settings.mjs]: The statement `path.join` seems obsolete if there is only a single parameter assigned => nothing to join

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: ahh yes, app was already declared there, electron was probably automatically declared by the IDE
since you recommended that for the `getPath`

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: you mean `app` cause `this.app` is seen as not defined (surprisingly) by the IDE

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [inline src/web/mjs/engine/Settings.mjs]: This should be done for the expected condition (`File does not Exist`) somewhere in the try block.
This catch block catches any exception, including errors like `Out of Memory`, `Permission Denied`, ...
calling `saveProfile` may lead to further re-throw errors.

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: For consistency: Either use `this._bookmarkOutpotPath()` instead of `Engine.Settings.bookmarkDirectory.value` or keep as is and remove `this._bookmarkOutpotPath()`.

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: For consistency: Either use `this._bookmarkOutpotPath()` instead of `Engine.Settings.bookmarkDirectory.value` or keep as is and remove `this._bookmarkOutpotPath()`.

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: May make this a `getter`, so it can be used like a member and not like a function

**Review by 27136773 — APPROVED** [review]
Some room left for improvements, but generally it should work like expected.

File: src/web/mjs/engine/Settings.mjs
Comment (hypoflex) [reply src/web/mjs/engine/Settings.mjs]: ahh alright, i figured it was for loading the settings only

**Comment by 27136773** [comment]
I will try to get the new deployment pipeline running before merging this change

File: src/web/mjs/engine/Settings.mjs
Comment (27136773) [reply src/web/mjs/engine/Settings.mjs]: Yes you IDE is correct, its local variable

File: src/web/mjs/engine/Storage.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Storage.mjs:728]: get and set in JS is new one for me, this should be correct to what you meant

File: src/web/mjs/engine/Storage.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Storage.mjs]: that should be `this._bookmarkOutputPath`

File: src/web/mjs/engine/Storage.mjs
Comment (hypoflex) [inline src/web/mjs/engine/Storage.mjs]: that should be `this._bookmarkOutputPath`

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: The getter is not static, so it needs to be accessed via instance (`this`), also the getter already returns the `value` property, simply `this._bookmarkOutputPath` will do the trick

File: src/web/mjs/engine/Storage.mjs
Comment (27136773) [inline src/web/mjs/engine/Storage.mjs]: The getter is not static, so it needs to be accessed via instance (`this`), also the getter already returns the `value` property, simply `this._bookmarkOutputPath` will do the trick

Comment (hypoflex) [reply src/web/mjs/engine/Storage.mjs]: wouldn't be better to have the getter return the object `Engine.Settings.bookmarkDirectory` instead of only the value? that way it has more potential use

Comment (27136773) [reply src/web/mjs/engine/Storage.mjs]: Its up to you to decide how this should be used

**Pull request merged**
